### PR TITLE
Create an explicit structure for `where` clauses

### PR DESCRIPTION
**Pull request by fitzgen**
This will make #247 easier to implement, so we can determine which type parameters are referenced in which where clauses.

r? @nikomatsakis

### CODE REVIEW

**Comment by fitzgen** [comment]
arg I was only running the `larlpop-test` test >.<

**Comment by fitzgen** [comment]
Note to self: need to support both `for<'a> &'a T: Whatever` and `T: for<'a> Fn('a)`

**Comment by fitzgen** [comment]
Fixed tests; haven't added `for<'a>` support.

**Comment by fitzgen** [comment]
r? @nikomatsakis 

I've got `for<'a>` stuff working now, also added a few tokenizer, parser, and integration tests.

File: lalrpop/src/parser/test.rs [hunk 1]
```diff
@@ -94,3 +94,59 @@ fn match_complex() {
...
+        "where T: IntoIterator<Item = usize>",
+        "where T: 'a",
+        "where 'a: 'b",
+        "where for<'a> &'a T: Debug",
+        "where T: for<'a> Flobbles<'a>",
+        "where T: FnMut(usize)",
+        "where T: FnMut(usize, bool)",
+        "where T: FnMut() -> bool",
+        "where T: for<'a> FnMut(&'a usize)",
+        "where T: Debug, U: Display",
```
Comment (nikomatsakis) [inline lalrpop/src/parser/test.rs]: nice

**Comment by nikomatsakis** [comment]
Sorry it took me a while. This is really well done.

**Pull request merged**